src/Storages/KeyDescription.cpp

```diff
@@ -151,6 +151,15 @@ KeyDescription KeyDescription::getSortingKeyFromAST(
             throw Exception(ErrorCodes::DATA_TYPE_CANNOT_BE_USED_IN_KEY,
                             "Column {} with type {} is not allowed in key expression, it's not comparable",
                             backQuote(result.sample_block.getByPosition(i).name), result.data_types.back()->getName());
+
+        auto check = [&](const IDataType & type)
```

Comment:
Is there a specific reason we want to prevent the usage of these types in keys even when the new settings are enabled?

Reply:
We have code related to order by keys/partition by keys/primary keys that heavily uses `Field` structure to compare values. And it doesn't work properly with `Variant` and `Dynamic` types because `Field` doesn't contain information about data type and comparison between values of different type doesn't work.